New QuantLib benchmark for machine comparison

amd-jadutoit opened the pull request:
Many thanks to @klausspanderen for help and guidance on this work.

QuantLib is one of the few open source production quality quant libraries. As such, platforms such as Phoronix and others have used the QuantLib benchmark as a measure of how well different CPUs run "financial workloads". This is only meaningful in as much the QL benchmark resembles an overnight risk run (this is the main computational workload that investment banks
select hardware for).

This PR modifies the benchmark so that it is easier to compare the performance of modern servers on overnight risk workloads.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/lballabio/QuantLib?pullRequest=1962) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/lballabio/QuantLib?pullRequest=1962) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: klausspanderen<br/>:x: amd-jadutoit<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/lballabio/QuantLib?pullRequest=1962) it.</sub>

klausspanderen commented:
@amd-jadutoit looking into the CI builds it seems that we have a couple of linker issues in the CI workflows, which we need to fix first.

amd-jadutoit commented:
> @amd-jadutoit looking into the CI builds it seems that we have a couple of linker issues in the CI workflows, which we need to fix first.

Yes I saw - I didn't realise there was an automake build system in addition to CMake.  I'm looking into that now, the Windows failures as well.  Will take me a few days probably.

lballabio commented:
Thanks!  No hurry—the next release will be in July.

lballabio commented on CMakeLists.txt:
```diff
@@ -39,12 +39,12 @@ set(QL_INSTALL_CMAKEDIR "lib/cmake/${PACKAGE_NAME}" CACHE STRING
 option(QL_BUILD_FUZZ_TEST_SUITE "Build fuzz test suite" OFF) 
 option(QL_ENABLE_OPENMP "Detect and use OpenMP" OFF)
 option(QL_ENABLE_PARALLEL_UNIT_TEST_RUNNER "Enable the parallel unit test runner" OFF)
+option(QL_PRINT_TEST_SUITE_RUNTIMES "Print the runtimes of each test in the test suite" OFF)
```
As a compilation flag, this is going to affect both test suite and benchmark, right?  It might make more sense to only print the test times during the benchmark run.  Maybe the timing could go in the Benchmark class?

amd-jadutoit replied:
This flag causes the runtime of each test in the test suite to be printed.  When I was constructing the new benchmark, I needed to get some idea of how long the various tests took to run.  So I introduced this flag and the other code in the [associated commit](https://github.com/lballabio/QuantLib/pull/1962/commits/5b9f300359bc9f5cd1a1b81bbee0b21afb35f93e) so that when I ran the entire test suite, I could see how long things take.

I thought this might be generally useful to others - it may very well not be.  If it's not, then I would suggest pulling it out (and the other code in that commit).  The benchmark has its own timing code, independently of this.

What are your thoughts? Is timing individual test cases a generally useful thing?

lballabio replied:
Timing single test cases is definitely useful (I'm doing it in [one of the CI builds](https://github.com/lballabio/QuantLib/blob/master/.github/workflows/test-times.yml) to check that we're not adding test cases that take too long) but it's also already supported by the Boost framework—if you run

    ./quantlib-test-suite --logger=JUNIT,warning:HRF,message

you'll get the usual output to console, plus an xml file with information on the single test cases.

amd-jadutoit replied:
Ah - I didn't know this.  In which case that commit of mine serves no purpose and I'll remove it.

amd-jadutoit replied:
I've removed the commit

amd-jadutoit commented:
@klausspanderen  @lballabio  I've pushed an updated PR
  * I've reverted the change in master CMakeLists.txt 
  * I've fixed the automake build path, both normal and unity.  Unity seems a little slow, but it does compile successfully on my system
  * I've set the default benchmark size to 3, which means just running the benchmark executable with no command line arguments will complete in a minute or two.

I'm not clear why the Windows CMake target was failing ... there is nothing Windows specific in the `test-suite` CMakeLists file.  I hope once the CI pipes have run this time I'll be able to get a hint as to what is wrong.

lballabio commented:
It looks like splitting the test suite into library + executable break the detection of the test cases?  I wouldn't be opposed to duplication here if that's what it takes...

amd-jadutoit commented:
@lballabio @klausspanderen I hope this will pass now
  * Automake now works my end with and without static libs configured.  It turns out that automake does not rebuild all the test object files when the same sources are passed to test-suite or benchmark, so this simplifies things considerably
  * On windows, there were a few unused vars in the benchmark program which were treated as errors.  The benchmark failed to build, which stopped the test suite for being built as well (I'm not sure why)
  * The final set of failures were due to the static library of test file objects.  It turns out that when you link a static lib of boost unit test objects against the boost unit test driver object, no undefined symbols are found and the static library is essentially ignored.  Hence, a working test executable with no tests.  This is fixed now (at least on my system).

lballabio commented:
Thanks—I added a small fix for "make dist" but it works now.

We don't have a CI build running the benchmark.  Do you think I should add one?

amd-jadutoit commented:
> Thanks—I added a small fix for "make dist" but it works now.
> 
> We don't have a CI build running the benchmark. Do you think I should add one?

Thanks very much for the fix! 

Yes please, I think that would be wise.  As I discovered, when making changes to the build system it's sometimes easy to mess something up, and then an executable might not get built.  Since the cost of building the benchmark is just a re-link of existing test-suite objects, it's negligible.

Running the benchmark with no arguments will run through all the 80+ tasks 3 times, which is pretty quick even on a small machine.  The process should exit successfully, `main()` should return 0.  If anything goes wrong, `main()` should return non-zero.

I don't think a separate CI pipe/build is necessary.  We could add this as another executable to run on some existing builds, provided we make sure we build the benchmark as well.  On CMake builds we should always build the benchmark if we build the test suite.  On automake I think we need to explicitly enable it.
